gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientGrpcDownscopingInterceptor.java

```diff
@@ -0,0 +1,226 @@
+import com.google.storage.v2.ComposeObjectRequest;
+import com.google.storage.v2.ComposeObjectRequest.SourceObject;
+import com.google.storage.v2.DeleteObjectRequest;
+import com.google.storage.v2.ReadObjectRequest;
+import com.google.storage.v2.StartResumableWriteRequest;
+import com.google.storage.v2.WriteObjectRequest;
+import com.google.storage.v2.WriteObjectSpec;
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ClientInterceptor;
+import io.grpc.ForwardingClientCall.SimpleForwardingClientCall;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Function;
+
+/**
+ * Interceptor to set the downscoped token.
+ *
+ * <p>When downscoping is enabled, the downscoped token is generated for each request. Java storage
+ * client currently does not have the capability to set this for each of the requests. As a work
+ * around the credentials will be set from the interceptor.
+ *
+ * <p>This is a short term solution till Java storage client adds support for setting credentials at
+ * a request level.
+ */
+class GoogleCloudStorageClientGrpcDownscopingInterceptor implements ClientInterceptor {
+
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+
+  @VisibleForTesting
+  static final String GOOGLE_STORAGE_V_2_STORAGE_COMPOSE_OBJECT =
+      "google.storage.v2.Storage/ComposeObject";
+
+  @VisibleForTesting
+  static final String GOOGLE_STORAGE_V_2_STORAGE_DELETE_OBJECT =
+      "google.storage.v2.Storage/DeleteObject";
+
+  @VisibleForTesting
+  static final String GOOGLE_STORAGE_V_2_STORAGE_READ_OBJECT =
+      "google.storage.v2.Storage/ReadObject";
+;
+
+  @VisibleForTesting
+  static final String GOOGLE_STORAGE_V_2_STORAGE_START_RESUMABLE_WRITE =
+      "google.storage.v2.Storage/StartResumableWrite";
+
+  @VisibleForTesting
+  static final String GOOGLE_STORAGE_V_2_STORAGE_WRITE_OBJECT =
+      "google.storage.v2.Storage/WriteObject";
+
+  @VisibleForTesting private static final int BUCKET_PREFIX_LENGTH = "projects/_/buckets/".length();
+
+  @VisibleForTesting
+  static Metadata.Key<String> AUTH_KEY =
+      Metadata.Key.of(AUTHORIZATION, Metadata.ASCII_STRING_MARSHALLER);
+
+  private final Function<List<AccessBoundary>, String> downscopingFunction;
+
+  public GoogleCloudStorageClientGrpcDownscopingInterceptor(
+      Function<List<AccessBoundary>, String> downscopedAccessToken) {
+    this.downscopingFunction = downscopedAccessToken;
+  }
+
+  @Override
+  public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+      MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+    logger.atFinest().log(String.format("interceptCall(): method=%s", method.getFullMethodName()));
+    return new SimpleForwardingClientCall<ReqT, RespT>(next.newCall(method, callOptions)) {
+      private int flowControlRequests;
+      private String methodName;
```

Comment:
```suggestion
      private final String methodName = method.genFullMethodName();
```

`start()` isn't a constructor, so it is a bit confusing initializing it there.

Reply:
Done